From backport, 396e0b39f320e31bead99a31f2284eabd62dd176
From: Bastian Germann <bage@debian.org>
Date: Mon, 29 May 2023 00:33:38 +0200
Subject: Exclude md5global.h

---
diff --git a/include/hmac-md5.h b/include/hmac-md5.h
index fb4a740c..7d84458a 100755
--- a/include/hmac-md5.h
+++ b/include/hmac-md5.h
@@ -17,8 +17,8 @@ typedef struct HMAC_MD5_CTX_s {
  *  values stored in network byte order (Big Endian)
  */
 typedef struct HMAC_MD5_STATE_s {
-    UINT4 istate[4];
-    UINT4 ostate[4];
+    uint32_t istate[4];
+    uint32_t ostate[4];
 } HMAC_MD5_STATE;
 
 #ifdef __cplusplus
diff --git a/lib/md5.c b/lib/md5.c
index e327679e..612cfa68 100644
--- a/lib/md5.c
+++ b/lib/md5.c
@@ -26,15 +26,16 @@ documentation and/or software.
 */
 
 #include <config.h>
-#include "md5global.h"
 #include "hmac-md5.h"
 
 #ifndef WIN32
 # include <arpa/inet.h>
 #endif
 
-static void MD5_memcpy PROTO_LIST ((POINTER, POINTER, unsigned int));
-static void MD5_memset PROTO_LIST ((POINTER, int, unsigned int));
+typedef unsigned char *POINTER;
+
+static void MD5_memcpy (POINTER, POINTER, unsigned int);
+static void MD5_memset (POINTER, int, unsigned int);
 
 /* Note: Replace "for loop" with standard memcpy if possible.
 
