Origin: upstream, 4798f8cae5cedbe5c53ae034e0bbca50896e9094
From: Bastian Germann <bage@debian.org>
Date: Tue, 30 May 2023 14:12:22 +0200
Subject: Replace custom with standard memcpy

Signed-off-by: Bastian Germann <bage@debian.org>
---
 lib/md5.c | 27 ++++-----------------------
 1 file changed, 4 insertions(+), 23 deletions(-)

diff --git a/lib/md5.c b/lib/md5.c
index a53fa7ff..9afc65c5 100644
--- a/lib/md5.c
+++ b/lib/md5.c
@@ -33,25 +33,6 @@ documentation and/or software.
 # include <arpa/inet.h>
 #endif
 
-typedef unsigned char *POINTER;
-
-static void MD5_memcpy (POINTER, POINTER, unsigned int);
-
-/* Note: Replace "for loop" with standard memcpy if possible.
-
-        */
-
-static void MD5_memcpy (output, input, len)
-POINTER output;
-POINTER input;
-unsigned int len;
-{
-       unsigned int i; 
-
-       for (i = 0; i < len; i++) 
-	      output[i] = input[i]; 
-}
-
 void _sasl_hmac_md5_init(HMAC_MD5_CTX *hmac,
 			 const unsigned char *key,
 			 int key_len)
@@ -91,8 +72,8 @@ void _sasl_hmac_md5_init(HMAC_MD5_CTX *hmac,
   /* start out by storing key in pads */
   OPENSSL_cleanse(k_ipad, sizeof(k_ipad));
   OPENSSL_cleanse(k_opad, sizeof(k_opad));
-  MD5_memcpy( k_ipad, (POINTER)key, key_len);
-  MD5_memcpy( k_opad, (POINTER)key, key_len);
+  memcpy(k_ipad, key, key_len);
+  memcpy(k_opad, key, key_len);
 
   /* XOR key with ipad and opad values */
   for (i=0; i<64; i++) {
@@ -218,8 +199,8 @@ unsigned char *digest; /* caller digest to be filled in */
   /* start out by storing key in pads */
   OPENSSL_cleanse(k_ipad, sizeof(k_ipad));
   OPENSSL_cleanse(k_opad, sizeof(k_opad));
-  MD5_memcpy( k_ipad, (POINTER)key, key_len);
-  MD5_memcpy( k_opad, (POINTER)key, key_len);
+  memcpy(k_ipad, key, key_len);
+  memcpy(k_opad, key, key_len);
 
   /* XOR key with ipad and opad values */
   for (i=0; i<64; i++) {
