From 8b1ba9cf1326acb88e4f236e043fad7d19f4f80a Mon Sep 17 00:00:00 2001
From: Jakub Jelen <jjelen@redhat.com>
Date: Thu, 1 Dec 2016 18:00:03 +0100
Subject: [PATCH] Missing definition of OPENSSL_zalloc for old OpenSSL versions

---
 saslauthd/lak.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/saslauthd/lak.c b/saslauthd/lak.c
index 53ed5ab9..6f1c7f3f 100644
--- a/saslauthd/lak.c
+++ b/saslauthd/lak.c
@@ -1716,6 +1716,15 @@ static int lak_check_password(
 #define EVP_MD_CTX_new()      EVP_MD_CTX_create()
 #define EVP_MD_CTX_free(ctx)  EVP_MD_CTX_destroy((ctx))
 
+static void *OPENSSL_zalloc(size_t num)
+{
+    void *ret = OPENSSL_malloc(num);
+
+    if (ret != NULL)
+        memset(ret, 0, num);
+    return ret;
+}
+
 static EVP_ENCODE_CTX *EVP_ENCODE_CTX_new(void)
 {
 	return OPENSSL_zalloc(sizeof(EVP_ENCODE_CTX));
