Index: os-cyrus-sasl2/plugins/ntlm.c
===================================================================
--- os-cyrus-sasl2.orig/plugins/ntlm.c
+++ os-cyrus-sasl2/plugins/ntlm.c
@@ -81,7 +81,6 @@
 #ifdef OPENSSL_NO_MD4
 #error No MD4 support in OpenSSL
 #endif
-#include <openssl/md5.h>
 #include <openssl/hmac.h>
 #include <openssl/des.h>
 #include <openssl/opensslv.h>
@@ -120,7 +119,6 @@ typedef unsigned char des_key_schedule[8
 #include "crypto-compat.h"
 
 #include <sasl.h>
-#define MD5_H  /* suppress internal MD5 */
 #include <saslplug.h>
 
 #include "plugin_common.h"
Index: os-cyrus-sasl2/plugins/otp.c
===================================================================
--- os-cyrus-sasl2.orig/plugins/otp.c
+++ os-cyrus-sasl2/plugins/otp.c
@@ -54,10 +54,8 @@
 #include <assert.h>
 
 #include <openssl/evp.h>
-#include <openssl/md5.h> /* XXX hack for OpenBSD/OpenSSL cruftiness */
 
 #include <sasl.h>
-#define MD5_H  /* suppress internal MD5 */
 #include <saslplug.h>
 
 #include "plugin_common.h"
Index: os-cyrus-sasl2/plugins/passdss.c
===================================================================
--- os-cyrus-sasl2.orig/plugins/passdss.c
+++ os-cyrus-sasl2/plugins/passdss.c
@@ -67,7 +67,6 @@
 /* for digest and cipher support */
 #include <openssl/evp.h>
 #include <openssl/hmac.h>
-#include <openssl/md5.h>
 #include <openssl/sha.h>
 #include <openssl/dsa.h>
 
@@ -75,7 +74,6 @@
 #include "crypto-compat.h"
 
 #include <sasl.h>
-#define MD5_H  /* suppress internal MD5 */
 #include <saslplug.h>
 
 #include "plugin_common.h"
Index: os-cyrus-sasl2/plugins/srp.c
===================================================================
--- os-cyrus-sasl2.orig/plugins/srp.c
+++ os-cyrus-sasl2/plugins/srp.c
@@ -93,7 +93,6 @@ typedef unsigned short uint32;
 #include "crypto-compat.h"
 
 #include <sasl.h>
-#define MD5_H  /* suppress internal MD5 */
 #include <saslplug.h>
 
 #include "plugin_common.h"
